File: JavaScript/1429-First Unique Number.js
/**
 * @param {number[]} nums
 */
var FirstUnique = function(nums) {
    this.hmap = new Map();
    this.queue = new Array();
    for (let num of nums){
        this.add(num);
    }
};

/**
 * @return {number}
 */
FirstUnique.prototype.showFirstUnique = function() {
    while (this.queue.length && this.hmap.get(this.queue[0]) === -1){
        this.queue.shift();
    }
    if (this.queue.length){
        return this.queue[0];
    }
    return -1;
};

/** 
 * @param {number} value
 * @return {void}
 */
FirstUnique.prototype.add = function(value) {
    if (this.hmap.has(value)){
        this.hmap.set(value, -1);
    }else{
        this.queue.push(value);
        this.hmap.set(value, 1);
    }
};

/** 
 * Your FirstUnique object will be instantiated and called as such:
 * var obj = new FirstUnique(nums)
 * var param_1 = obj.showFirstUnique()
 * obj.add(value)
 */

File: C++/1429-First Unique Number.cpp
class FirstUnique {
public:
    unordered_map<int, int> count;
    deque<int> dq;
    FirstUnique(vector<int>& nums) {
        for (int num:nums) {
            count[num]++;
            dq.push_back(num);
        }
    }
    
    int showFirstUnique() {
        while (!dq.empty() && count[dq.front()] > 1) {
            dq.pop_front();
        }
        if (dq.empty()) return -1;
        return dq.front();
    }
    
    void add(int value) {
        count[value]++;
        dq.push_back(value);
    }
};

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique* obj = new FirstUnique(nums);
 * int param_1 = obj->showFirstUnique();
 * obj->add(value);
 */



// Use list (not faster)
class FirstUnique {
public:
    list<int> ls;  // store unique values
    unordered_map<int, list<int>::iterator> hmap;
    FirstUnique(vector<int>& nums) {
        for (int num:nums) {
            add(num);
        }
    }
    
    int showFirstUnique() {
        if (!ls.empty()) {
            return ls.front();
        }
        return -1;
    }
    
    void add(int value) {
        if (!hmap.contains(value)) {
            ls.push_back(value);
            hmap[value] = --ls.end();
        } else if (hmap[value] != ls.end()) {
            ls.erase(hmap[value]);
            hmap[value] = ls.end();
        }
    }
};

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique* obj = new FirstUnique(nums);
 * int param_1 = obj->showFirstUnique();
 * obj->add(value);
 */